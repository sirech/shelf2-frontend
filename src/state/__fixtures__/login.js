const login = () => ({
  authenticated: true,
  failed: false,
})

export default login
