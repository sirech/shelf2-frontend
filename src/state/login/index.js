import * as constants from './constants'
import * as actions from './actions'

export { constants, actions }
export { default as loginSelector } from './selectors'
export { default as reducer } from './reducers'
