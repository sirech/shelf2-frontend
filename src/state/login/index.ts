import * as constants from './constants'
import * as actions from './actions'

export { constants, actions }
export {
  default as loginSelector,
  authenticatedSelector,
  failedSelector,
} from './selectors'
export { default as reducer } from './reducers'
