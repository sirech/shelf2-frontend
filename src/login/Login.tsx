import React, { useEffect } from 'react'

import { Redirect } from 'react-router-dom'
import { useAppDispatch, useAppSelector } from 'hooks'

import { actions, authenticatedSelector } from 'state/login'

const redirectIfLoggedIn = (authenticated: boolean) => {
  if (authenticated) {
    return <Redirect to="/" />
  }
}
const Login = () => {
  const authenticated = useAppSelector(authenticatedSelector)
  const dispatch = useAppDispatch()

  useEffect(() => {
    if (!authenticated) {
      dispatch(actions.startLogin())
    }
  }, [authenticated, dispatch])

  return <div>{redirectIfLoggedIn(authenticated)}</div>
}

export default Login
