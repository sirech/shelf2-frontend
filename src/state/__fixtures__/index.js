import R from 'ramda'

import books from './books'
import years from './years'
import form from './form'
import login from './login'

const modal = () => ({
  opened: false
})

const search = R.pipe(books, R.omit(['activeYear']))

const state = () => {
  return {
    books: books(),
    search: search(),
    years: years(),
    form: form(),
    modal: modal(),
    login: login()
  }
}

export { books, search, years, form, login }
export default state
