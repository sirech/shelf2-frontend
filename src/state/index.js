import { combineReducers } from 'redux'

import * as R from 'ramda'

import { constants as booksConstants, reducer as booksReducer } from './books'
import {
  constants as searchConstants,
  reducer as searchReducer,
} from './search'
import { constants as yearsConstants, reducer as yearsReducer } from './years'
import { constants as formConstants, reducer as formReducer } from './form'
import { constants as modalConstants, reducer as modalReducer } from './modal'
import { constants as loginConstants, reducer as loginReducer } from './login'

export const actionPicker = list => {
  return actions => {
    const h = R.pick(list)(actions)

    if (Object.keys(h).length !== list.length) {
      throw new Error(
        `Not all requested actions (${list}) exist in the given actions`
      )
    }

    return h
  }
}

const rootReducer = combineReducers({
  [booksConstants.namespace]: booksReducer,
  [searchConstants.namespace]: searchReducer,
  [yearsConstants.namespace]: yearsReducer,
  [formConstants.namespace]: formReducer,
  [modalConstants.namespace]: modalReducer,
  [loginConstants.namespace]: loginReducer,
})

export default rootReducer
