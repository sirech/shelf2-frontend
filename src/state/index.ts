import { ThunkAction, UnknownAction, configureStore } from '@reduxjs/toolkit'

import { constants as booksConstants, reducer as booksReducer } from './books'
import {
  constants as searchConstants,
  reducer as searchReducer,
} from './search'
import { constants as yearsConstants, reducer as yearsReducer } from './years'
import { constants as loginConstants, reducer as loginReducer } from './login'

const store = configureStore({
  reducer: {
    [booksConstants.namespace]: booksReducer,
    [searchConstants.namespace]: searchReducer,
    [yearsConstants.namespace]: yearsReducer,
    [loginConstants.namespace]: loginReducer,
  },
})

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  UnknownAction
>

export default store
