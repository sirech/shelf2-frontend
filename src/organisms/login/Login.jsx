// @flow

import React from 'react'
import { connect } from 'react-redux'

import { createStructuredSelector } from 'reselect'
import { Redirect } from 'react-router-dom'

import GoogleLogin from 'react-google-login'

import { Col, Row, Alert } from 'reactstrap'

import { actionPicker } from 'state'
import { actions, authenticatedSelector, failedSelector } from 'state/login'

type Props = {
  authenticated: boolean,
  failed: boolean,
  login: string => undefined,
  loginFailure: _ => undefined,
}

// exported for testing
export class Login extends React.Component {
  static defaultProps: Props
  props: Props
  onSuccess: Function
  onFailure: Function

  constructor() {
    super()
    this.onSuccess = this.onSuccess.bind(this)
    this.onFailure = this.onFailure.bind(this)
  }

  redirectIfLoggedIn() {
    if (this.props.authenticated) {
      return <Redirect to="/" />
    }
  }

  errorMessage() {
    if (this.props.failed) {
      return <Alert color="danger">Could not log in</Alert>
    }
  }

  onSuccess(response: { accessToken: string }) {
    this.props.login(response.accessToken)
  }

  onFailure(_) {
    this.props.loginFailure()
  }

  render() {
    return (
      <Row>
        {this.redirectIfLoggedIn()}
        <Col xs="12">
          {this.errorMessage()}
          <div className="text-center">
            <GoogleLogin
              clientId="821590472100-8l14rsm3sof4so0p495tkaf6lgd5p47s.apps.googleusercontent.com"
              buttonText="Login with Google"
              onSuccess={this.onSuccess}
              onFailure={this.onFailure}
            />
          </div>
        </Col>
      </Row>
    )
  }
}

Login.defaultProps = {
  authenticated: false,
  failed: false,
  login: _ => undefined,
  loginFailure: _ => undefined,
}

export default connect(
  (state, props) =>
    createStructuredSelector({
      authenticated: authenticatedSelector,
      failed: failedSelector,
    })(state),
  actionPicker(['login', 'loginFailure'])(actions)
)(Login)
