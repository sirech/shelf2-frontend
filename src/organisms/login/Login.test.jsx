import React from 'react'
import { Route } from 'react-router-dom'

import { fullRender } from 'test'
import state from 'state/__fixtures__'

import Login from './Login'

import { actions } from 'state/login'
jest.mock('state/login/actions', () => ({
  startLogin: jest.fn((dispatch) => jest.fn()),
}))

describe('components', () => {
  describe('Login', () => {
    it('does not blow up', () => {
      const currentState = state()
      currentState.login.authenticated = false
      fullRender(
        <Route path="/login" component={Login} />,
        currentState,
        '/login'
      )
      expect(actions.startLogin.mock.calls.length).toBe(1)
    })

    it('renders a redirect if authenticated', () => {
      const currentState = state()
      currentState.login.authenticated = true
      const { component } = fullRender(
        <Route path="/login" component={Login} />,
        currentState,
        '/login'
      )
      expect(component.toJSON()).toBeNull()
    })
  })
})
