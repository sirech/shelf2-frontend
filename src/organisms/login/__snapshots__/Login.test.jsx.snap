// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components Login does not blow up 1`] = `
<Row
  tag="div"
>
  <Col
    tag="div"
    widths={
      Array [
        "xs",
        "sm",
        "md",
        "lg",
        "xl",
      ]
    }
    xs="12"
  >
    <div
      className="text-center"
    >
      <e
        accessType="online"
        buttonText="Login with Google"
        clientId="821590472100-8l14rsm3sof4so0p495tkaf6lgd5p47s.apps.googleusercontent.com"
        cookiePolicy="single_host_origin"
        disabledStyle={
          Object {
            "opacity": 0.6,
          }
        }
        fetchBasicProfile={true}
        icon={true}
        isSignedIn={false}
        jsSrc="https://apis.google.com/js/api.js"
        onFailure={[Function]}
        onRequest={[Function]}
        onSuccess={[Function]}
        prompt=""
        scope="profile email"
        tag="button"
        theme="light"
        type="button"
        uxMode="popup"
      />
    </div>
  </Col>
</Row>
`;
